### PR TITLE
[Runtime] SE-0143: Evaluate conditional conformances at runtime.

### PR DESCRIPTION
**Pull request by DougGregor**
When evaluating whether a given type conforms to a protocol, evaluate the
conditional requirements and pass the results to the witness table
accessor function. This provides the ability to query conditional
conformances at runtime, and is the last major part of implementing
SE-0143.

The newly-added unlock/lock dance in the conformance lookup code is a
temporary stub. We have some ideas to do this better.

Fixes rdar://problem/34944655.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Review by gparker42 — DISMISSED** [review]
TwoWordPair is going away in #14373. This should wait until that merges.

**Comment by DougGregor** [comment]
Bah, it's going to conflict with #14373 even once I've reverted the first commit. I'll rebase once that goes in.

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

File: CHANGELOG.md [hunk 1]
```diff
@@ -29,6 +29,12 @@ Swift 5.0
   circumstances with very simple values, so this is unlikely to affect
   real-world code.)
 
+* [SE-0143][]
+
+	Runtime query of conditional conformances is now implemented. Therefore,
+	a dynamic cast such as `value as? P`, where the dynamic type of `value`
+	conditional conforms to `P`, will succeed when the conditional
```
Comment (jckarter) [inline CHANGELOG.md]: \*conditional**ly**

Comment (DougGregor) [reply CHANGELOG.md]: Oops, thanks. I'll fix that in a follow-up commit.

**Comment by pitiphong-p** [comment]
Will this `Evaluate conditional conformances at runtime` get into Swift 4.1?

**Comment by jckarter** [comment]
No.

**Comment by DougGregor** [comment]
@pitiphong-p : No, this will not go into Swift 4.1. It's dependent on a large number of changes in the Swift compiler and runtime that would not be safe to back port to Swift 4.1.

**Comment by pitiphong-p** [comment]
I see. Thank you.

**Pull request merged**